src/Elf.h

```diff
@@ -0,0 +1,395 @@
+#include <memory>
+#include <vector>
+#include <list>
+#include <string>
+#include <iterator>
+#include <limits>
+
+namespace Halide {
+namespace Internal {
+namespace Elf {
+
+// This ELF parser mostly deserializes the object into a graph
+// structure in memory. It replaces indices into tables (sections,
+// symbols, etc.) with a weakly referenced graph of pointers. The
+// Object datastructure owns all of the objects. This namespace exists
+// because it is very difficult to use LLVM's object parser to modify
+// an object (it's fine for parsing only).
+
+class Object;
+class Symbol;
+class Section;
+class Relocation;
+
+// Helpful wrapper to allow range-based for loops.
+template <typename T>
+class iterator_range {
+    T b, e;
+public:
+    iterator_range(T b, T e) : b(b), e(e) {}
+
+    T begin() const { return b; }
+    T end() const { return e; }
+};
+
+/** Describes a symbol */
+class Symbol {
+public:
+    enum Binding : uint8_t {
+        STB_LOCAL = 0,
+        STB_GLOBAL = 1,
+        STB_WEAK = 2,
+        STB_LOPROC = 13,
+        STB_HIPROC = 15,
+    };
+
+    enum Type : uint8_t {
+        STT_NOTYPE = 0,
+        STT_OBJECT = 1,
+        STT_FUNC = 2,
+        STT_SECTION = 3,
+        STT_FILE = 4,
+        STT_LOPROC = 13,
+        STT_HIPROC = 15,
+    };
+
+    enum Visibility : uint8_t {
+        STV_DEFAULT = 0,
+        STV_INTERNAL = 1,
+        STV_HIDDEN = 2,
+        STV_PROTECTED = 3,
+    };
+
+private:
+    std::string name;
+    const Section *definition = nullptr;
+    uint64_t offset = 0;
+    uint32_t size = 0;
+    Binding binding = STB_LOCAL;
+    Type type = STT_NOTYPE;
+    Visibility visibility = STV_DEFAULT;
+
+public:
+    Symbol() {}
+    Symbol(const std::string &name) : name(name) {}
+
+    /** Accesses the name of this symbol. */
+    ///@{
+    Symbol &set_name(const std::string &name) { this->name = name; return *this; }
+    const std::string &get_name() const { return name; }
+    ///@}
+
+    /** Accesses the type of this symbol. */
+    ///@{
+    Symbol &set_type(Type type) { this->type = type; return *this; }
+    Type get_type() const { return type; }
+    ///@}
+
+    /** Accesses the properties that describe the definition of this symbol. */
+    ///@{
+    Symbol &define(const Section *section, uint64_t offset, uint32_t size) {
+        this->definition = section;
+        this->offset = offset;
+        this->size = size;
+        return *this;
+    }
+    bool is_defined() const { return definition != nullptr; }
+    const Section *get_section() const { return definition; }
+    uint64_t get_offset() const { return offset; }
+    uint32_t get_size() const { return size; }
+    ///@}
+
+    /** Access the binding and visibility of this symbol. See the ELF
+     * spec for more information about these properties. */
+    ///@{
+    Symbol &set_binding(Binding binding) {
+        this->binding = binding;
+        return *this;
+    }
+    Symbol &set_visibility(Visibility visibility) {
```

Comment:
I made all of the multi-statement setters multi-line, left the single statement things (getters) single line.